### PR TITLE
onepad 1.3: import lilypad state machine into onepad

### PR DESCRIPTION
**Pull request by gregory38**
Not tested
* 2 pads (tested Keyboard + pad)
* load/save state

Require a ton of linux testing

### CODE REVIEW

**Comment by kust2708** [comment]
I have a page fault each time I try to launch a game.

```
/media/Data/Documents/Autres/pcsx2/common/src/Utilities/Linux/LnxHostSys.cpp(70) : assertion failed:
    Function:  void SysPageFaultSignalFilter(int, siginfo_t*, void*)
    Thread:    Unknown
    Condition: false
    Message:   Unhandled page fault @ 0x00000000

Stacktrace:
[00] 0x0x5699eb32                                
[01] 0x0x5699ec9d                                
[02] 0x0x569adaaa                                
[03] __kernel_rt_sigreturn                       
[04] 0x0xf604ecc0                                
[05] 0x0xf6046d58                                
[06] 0x0x569adfd7                                
[07] 0x0x5688b8c4                                
[08] 0x0x5662e29f                                
[09] 0x0x569c2049                                
[10] 0x0x569c22fd                                
[11] 0x0x569c234a                                
[12] 0x0x569c4762                                
[13] 0x0xf6043301                                
[14] clone                                       
```

**Comment by gregory38** [comment]
@kust2708 
Could you retry with a dev (or a prof build) so I have a nice backtrace (symbol + no omit frame pointer). Or use gdb.

By the way, I think we have a small issue with the GUI.
There are a mix of ID for the rumble detection GUI will use the tab number (0 or 1).
```
int gamepad_id = m_tab_gamepad->GetSelection(); // get the tab ID (equivalent to the gamepad id)
 GamepadConfiguration gamepad_config(gamepad_id, this);
```
But the class requires the index of the gamepad array. 
```
void GamepadConfiguration::InitGamepadConfiguration()
     if(!s_vgamePad[m_pad_id]->TestForce(0.001f))
```

**Comment by kust2708** [comment]
It's was already compiled with the debug flag, here is the trace of gdb:
```
(gdb) run
Starting program: /media/Data/Documents/Autres/pcsx2/bin/PCSX2 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
11:37:16: Warning: Mismatch between the program and library build versions detected.
The library used 3.0 (wchar_t,compiler with C++ ABI 1009,wx containers,compatible with 2.8),
and your program used 3.0 (wchar_t,compiler with C++ ABI 1010,wx containers,compatible with 2.8).

(PCSX2:8077): Gtk-WARNING **: Unable to locate theme engine in module_path: "adwaita",

(PCSX2:8077): Gtk-WARNING **: Unable to locate theme engine in module_path: "adwaita",
Interface is initializing.  Entering Pcsx2App::OnInit!
Applying operating system default language...
Loading language translation databases for 'French' [fr_FR]
Command line parsing...
Command line parsed!
[New Thread 0xf3fe5b40 (LWP 8091)]
(UserMode) Found portable install ini @ /media/Data/Documents/Autres/pcsx2/bin/portable.ini
[New Thread 0xf35ffb40 (LWP 8092)]
[New Thread 0xf2bffb40 (LWP 8093)]
[New Thread 0xf21ffb40 (LWP 8096)]
[New Thread 0xf15c9b40 (LWP 8097)]
[Thread 0xf15c9b40 (LWP 8097) exited]
[New Thread 0xf15c9b40 (LWP 8098)]
[New Thread 0xee1adb40 (LWP 8099)]
[New Thread 0xe03bab40 (LWP 8101)]
[Thread 0xe03bab40 (LWP 8101) exited]
[New Thread 0xe03bab40 (LWP 8102)]
[Thread 0xe03bab40 (LWP 8102) exited]
[New Thread 0xe03bab40 (LWP 8103)]
[Thread 0xe03bab40 (LWP 8103) exited]
[New Thread 0xe03bab40 (LWP 8104)]
[Thread 0xe03bab40 (LWP 8104) exited]
[New Thread 0xe03bab40 (LWP 8105)]
[Thread 0xe03bab40 (LWP 8105) exited]
[New Thread 0xe03bab40 (LWP 8107)]
[New Thread 0xdb9feb40 (LWP 8108)]
[Thread 0xdb9feb40 (LWP 8108) exited]
[Thread 0xe03bab40 (LWP 8107) exited]

Thread 8 "MTGS" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xee1adb40 (LWP 8099)]
0xf6854cc0 in __lll_unlock_elision () from /usr/lib32/libpthread.so.0
```

**Comment by gregory38** [comment]
@kust2708 did you miss some part of the copy past like the back trace? 

By the way, this is very bad warning that could explain your issue. Wx and PCSX2 wasn't compiled with the same compiler ABI. It could explain some crashes.
```
11:37:16: Warning: Mismatch between the program and library build versions detected.
The library used 3.0 (wchar_t,compiler with C++ ABI 1009,wx containers,compatible with 2.8),
and your program used 3.0 (wchar_t,compiler with C++ ABI 1010,wx containers,compatible with 2.8).
```

**Comment by kust2708** [comment]
Yes, I forgot the backtrace. It doesn't seems to be related to your work on onepad.
```
(gdb) run
Starting program: /media/Data/Documents/Autres/pcsx2/bin/PCSX2 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
12:05:41: Warning: Mismatch between the program and library build versions detected.
The library used 3.0 (wchar_t,compiler with C++ ABI 1009,wx containers,compatible with 2.8),
and your program used 3.0 (wchar_t,compiler with C++ ABI 1010,wx containers,compatible with 2.8).

(PCSX2:15703): Gtk-WARNING **: Unable to locate theme engine in module_path: "adwaita",

(PCSX2:15703): Gtk-WARNING **: Unable to locate theme engine in module_path: "adwaita",
Interface is initializing.  Entering Pcsx2App::OnInit!
Applying operating system default language...
Loading language translation databases for 'French' [fr_FR]
Command line parsing...
Command line parsed!
[New Thread 0xf3fe5b40 (LWP 15717)]
(UserMode) Found portable install ini @ /media/Data/Documents/Autres/pcsx2/bin/portable.ini
[New Thread 0xf35ffb40 (LWP 15718)]
[New Thread 0xf2bffb40 (LWP 15719)]
[New Thread 0xf21ffb40 (LWP 15722)]
[New Thread 0xf15c9b40 (LWP 15723)]
[Thread 0xf15c9b40 (LWP 15723) exited]
[New Thread 0xf15c9b40 (LWP 15737)]
[New Thread 0xee1aab40 (LWP 15738)]
[New Thread 0xdff7db40 (LWP 15739)]
[Thread 0xdff7db40 (LWP 15739) exited]
[New Thread 0xdff7db40 (LWP 15740)]
[Thread 0xdff7db40 (LWP 15740) exited]
[New Thread 0xdff7db40 (LWP 15741)]
[Thread 0xdff7db40 (LWP 15741) exited]
[New Thread 0xdff7db40 (LWP 15742)]
[Thread 0xdff7db40 (LWP 15742) exited]
[New Thread 0xdff7db40 (LWP 15743)]
[Thread 0xdff7db40 (LWP 15743) exited]
[New Thread 0xdff7db40 (LWP 15745)]
[New Thread 0xdb5feb40 (LWP 15746)]
[Thread 0xdb5feb40 (LWP 15746) exited]
[Thread 0xdff7db40 (LWP 15745) exited]

Thread 8 "MTGS" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xee1aab40 (LWP 15738)]
0xf6854cc0 in __lll_unlock_elision () from /usr/lib32/libpthread.so.0
(gdb) back
#0  0xf6854cc0 in __lll_unlock_elision () from /usr/lib32/libpthread.so.0
#1  0xf684cd58 in __pthread_mutex_unlock_usercnt () from /usr/lib32/libpthread.so.0
#2  0x569b1d7b in Threading::Mutex::Release (this=0x584b6954 <mtgsThread+340>) at /media/Data/Documents/Autres/pcsx2/common/src/Utilities/Mutex.cpp:184
#3  0x5688d874 in SysThreadBase::OnCleanupInThread (this=0x584b6800 <mtgsThread>) at /media/Data/Documents/Autres/pcsx2/pcsx2/System/SysThreadBase.cpp:253
#4  0x566088c1 in SysMtgsThread::OnCleanupInThread (this=0x584b6800 <mtgsThread>) at /media/Data/Documents/Autres/pcsx2/pcsx2/MTGS.cpp:598
#5  0x569c5a99 in Threading::pxThread::_try_virtual_invoke (this=0x584b6800 <mtgsThread>, method=&virtual Threading::pxThread::OnCleanupInThread()) at /media/Data/Documents/Autres/pcsx2/common/src/Utilities/ThreadTools.cpp:545
#6  0x569c5d65 in Threading::pxThread::_ThreadCleanup (this=0x584b6800 <mtgsThread>) at /media/Data/Documents/Autres/pcsx2/common/src/Utilities/ThreadTools.cpp:597
#7  0x569c820a in Threading::pxThread::_pt_callback_cleanup (handle=0x584b6800 <mtgsThread>) at /media/Data/Documents/Autres/pcsx2/common/src/Utilities/ThreadTools.cpp:169
#8  __pthread_cleanup_class::~__pthread_cleanup_class (this=<synthetic pointer>, __in_chrg=<optimized out>) at /usr/include/pthread.h:561
#9  Threading::pxThread::_internal_callback (itsme=0x584b6800 <mtgsThread>) at /media/Data/Documents/Autres/pcsx2/common/src/Utilities/ThreadTools.cpp:683
#10 0xf6849301 in start_thread () from /usr/lib32/libpthread.so.0
#11 0xf677490e in clone () from /usr/lib32/libc.so.6
```

**Comment by gregory38** [comment]
Yes someone post a similar issue somewhere ^^ I searched a bit, it is this one #1455 

I didn't manage to reproduce it on my side. I don't know if it is a regression, if yes when it was introduced. If you could find a bad commit, it would be very helpful.

**Comment by gregory38** [comment]
Except the initial bug report was during the exit not the start!

**Comment by kust2708** [comment]
Ah yes I get this bug on my last release build I did. I will search for a solution on my side.
I will continue to search for bugs as well.

**Comment by gregory38** [comment]
```
 I will continue to search for bugs as well. 
```
I won't be against a bug fix too :) You could like at rumble on this branch, I'm sure that I broke something. I don't know if rumble ever work on my side (my last test was done a very long time ago).

By the way, what is your distribution? 
```
The library used 3.0 (wchar_t,compiler with C++ ABI 1009,wx containers,compatible with 2.8),
and your program used 3.0 (wchar_t,compiler with C++ ABI 1010,wx containers,compatible with 2.8)
```
I think bad people add a -fabi-version=9 in wx compilation script. However GCC6 uses 10 by default.

**Comment by kust2708** [comment]
It was working on the GUI with my DS4. So I suppose that it works as well in game (I can't check for the moment).

**Comment by gregory38** [comment]
GUI isn't happy when you have multiple pads (3), with at least one that don't support rumble. Anyway, I changed all the state machine of the pad so the rumble function isn't call at the same place (and handling is different). Hopefully it is fine but without test no guarantee.

**Comment by kust2708** [comment]
Oh that's true. If you have at least one gamepad without rumble support, the GUI will lock rumble's setting for all of them. I didn't thought about this possibility.

File: pcsx2/SaveState.h [hunk 1]
```diff
@@ -24,7 +24,7 @@
 //  the lower 16 bit value.  IF the change is breaking of all compatibility with old
 //  states, increment the upper 16 bit value, and clear the lower 16 bits to 0.
 
-static const u32 g_SaveVersion = (0x9A0B << 16) | 0x0000;
+static const u32 g_SaveVersion = (0x9A0C << 16) | 0x0000;
```
Comment (turtleli) [inline pcsx2/SaveState.h]: I don't think you need to bump this save state version unless you've actually changed the save state format of PCSX2 itself.

**Comment by gregory38** [comment]
Still no news on the save state. Let's merge it and see if someone complains.

**Pull request merged**